proposals/NNNN-swiftpm-package-traits.md

```diff
@@ -0,0 +1,478 @@
+# Package traits
```

Comment:
I leave it to community wide bike shedding :D

Reply:
+1 on features

Reply:
Another argument against is searchability in documentation, imagine googling for "swift language feature" or "swiftpm feature" and trying to find relevant information on this, which would be basically impossible with this naming.

Reply:
Those are good points Max, I retract my +1 for features.

Reply:
On the other hand, Python refers to them as "optional features", which is eminently searchable.

Reply:
I am still leaning towards `traits` just for the fact that the `PackageDescription` already contains `.enableUpcomingFeature` which is tied to the compiler features and having two different kinds of feature already in the `PackageDescription` seems very confusing.

Reply:
Package flags, package options.

Reply:
Flags put us into a corner of making them binary: enabled/disabled. One nice direction I anticipate is codifying traits in a package as `RawRepresentable<String>` enums, also allowing enums with associated values. Then one could create a trait with `case stackSize(UInt)` encoded as `stackSize=42` trait for a stack size setting passed to a linker that can be useful in embedded use cases, for example. IMO from that perspective this looks even more like "package parameters" to me.

Reply:
Just for the record, the reason I dislike _traits_ is that I don't think this would be using the word in its usual (plain English) sense. Dictionary says:

> **trait** | treɪt, treɪ |
> noun
> a distinguishing quality or characteristic, typically one belonging to a person: _the traditionally British trait of self-denigration._
> * a genetically determined characteristic: _breeders were installing some trait that allowed the crop to thrive._

and I don't think these are distinguishing qualities or characteristics of the package. Rather, they're options that can be set.

"Options" is quite a nice name for that, actually. I do quite like that one.

Reply:
I'd like to toss "configuration" into the pool for discussion for good measure, though I wouldn't really be against "options".